### PR TITLE
Load & Store instructions

### PR DESCRIPTION
**Pull request by toon-leader-bacon**
Implementation and tests for the load and store instructions. 

Review, understand and merge the Proof of Concept (PoC) mr first: https://github.com/akm303/MIPS_Simulator/pull/9

### CODE REVIEW

**Review by akm303 — COMMENTED** [review]
same comments as previously for the most part, load store instructions look good on first glance, but hesitant to approve just on the [not knowing how github operates and it scares me, push the PoC through first maybe]

File: src/main/java/group1/mips_simulator/SimController.java [hunk 1]
```diff
@@ -133,7 +133,7 @@ public void loadPlusOnClick(ActionEvent actionEvent) {
     public void storeOnClick(ActionEvent actionEvent) {
         Register memBuffReg = this.computer.cpu.regfile.getMBR();
         Register memAddReg = this.computer.cpu.regfile.getMAR();
-        this.computer.mem.write(memAddReg.read(), memBuffReg.read());
+        this.computer.memory.write(memAddReg.read(), memBuffReg.read());
```
[the remark this answers is not in the document]
Comment (toon-leader-bacon) [reply src/main/java/group1/mips_simulator/SimController.java:136]: Updated. Good catch

File: src/main/java/group1/mips_simulator/components/Computer.java [hunk 2]
```diff
@@ -32,46 +33,194 @@ public Computer() {
...
+            // Floating Point Instructions/ Vector Operations
+            // TODO
+            default ->
+                    throw new IllegalArgumentException("Unknown instruction op code name: " + instruction.opCode.name +
+                            "\nInstruction: " + instruction);
+        }
+        // After every instruction then we increment the Program Counter
+        // EXCEPT if the instruction is a transfer instruction. Transfer instructions
+        // are responsible for setting the PC themselves.
+        if (!instruction.isTransferInstruction()) {
```
[the remark this answers is not in the document]
Comment (toon-leader-bacon) [reply src/main/java/group1/mips_simulator/components/Computer.java:105]: We discussed in person and I agree: These execute instructions should return a boolean representing if the computer should halt, AND the next program counter value

File: src/main/java/group1/mips_simulator/components/Computer.java [hunk 3]
```diff
@@ -32,46 +33,194 @@ public Computer() {
...
+     * - Indirection Bit
+     * - Index Registers
+     * <p>
+     * The specific instructions for calculating the EA are from the project documentation pg13
+     *
+     * @return The memory location that is the Effective Address for the provided
+     * instruction.
+     */
+    public Value calculateEA(RXIA_Instruction instruction) {
+        return this.calculateEA(instruction.getIX(), instruction.getAddress(), instruction.getI());
```
[the remark this answers is not in the document]
Comment (toon-leader-bacon) [reply src/main/java/group1/mips_simulator/components/Computer.java:136]: Agreed, updated.

[the remark this answers is not in the document]
Comment (toon-leader-bacon) [reply src/main/java/group1/mips_simulator/components/Computer.java:136]: I'm not sure what you're suggesting here. Please clarify if there's anything I need to do in this MR :)

File: src/main/java/group1/mips_simulator/components/ExecutionInstructions.java [hunk 4]
```diff
@@ -0,0 +1,268 @@
...
+
+import group1.mips_simulator.components.cpuParts.ConditionCode;
+import group1.mips_simulator.components.cpuParts.Register;
+import group1.mips_simulator.components.instructionParts.Field;
+import group1.mips_simulator.components.instructionParts.RXIA_Instruction;
+
+/**
+ * A class that holds all the functions for actually executing an Instruction object
+ */
+public class ExecutionInstructions {
```
[the remark this answers is not in the document]
Comment (toon-leader-bacon) [reply src/main/java/group1/mips_simulator/components/ExecutionInstructions.java:11]: Sure, that's fine. I just needed a place to put all these functions. I was thinking of making a dedicated "Execute_Blah" class for every type of instruction (`Execute_Jump` `Execute_Arithmetic` `Execute_Transfer` ect) but I remember you said you didn't like that our Assembler had so many classes. So I decided to smash them all into one `Execute_Generic_Instruction` type class. 

I don't really care about the name. Will refactor to change the name now 👍

File: src/main/java/group1/mips_simulator/components/ExecutionInstructions.java [hunk 5]
```diff
@@ -0,0 +1,268 @@
...
+    /**
+     * L0DR r, x, address[,I]
+     * opcode 01(octal)
+     * Load Register From Memory, r = 0..3
+     * r <− c(EA)
+     * note that EA is computed as given above
+     */
+    public void execute_ldr(Computer computer, RXIA_Instruction i) {
+        // r <- c(EA)
+        Value ea = computer.calculateEA(i);
```
[the remark this answers is not in the document]
Comment (toon-leader-bacon) [reply src/main/java/group1/mips_simulator/components/ExecutionInstructions.java:22]: 🤷  I like classes

File: src/main/java/group1/mips_simulator/components/cpuParts/RegisterFile.java [hunk 6]
```diff
@@ -3,32 +3,41 @@
...
+ * That's why we're using a fixed array for the GPR and IXR representations.
+ */
 public class RegisterFile {
 
     public static final int GPR_COUNT = 4;
     public static final int IXR_COUNT = 3;
 
-    /*
-    A register file is a collection of all registers accessible and usable by a CPU
+    public Register[] gpr = new Register[]{ //General Purpose Reg: R0-R3
```
[the remark this answers is not in the document]
Comment (toon-leader-bacon) [reply src/main/java/group1/mips_simulator/components/cpuParts/RegisterFile.java:17]: Ok, I don't really care I suppose. I'll try to revert it.

**Review by akm303 — APPROVED** [review]
looks good, let a few minor comments

File: src/main/java/group1/mips_simulator/components/cpuParts/RegisterFile.java [hunk 7]
```diff
@@ -10,18 +10,10 @@
...
-            new Register(), // GPR2
-            new Register()  // GPR3
-    };
-    public Register[] ixr = new Register[]{ //Index Reg: X1-X3.
-            new Register(), // IXR0
-            new Register(), // IXR1
-            new Register()  // IXR2
-    };
+    public static final int GPR_COUNT = 4;
+    public static final int IXR_COUNT = 3;
```
[the remark this answers is not in the document]
Comment (toon-leader-bacon) [reply src/main/java/group1/mips_simulator/components/cpuParts/RegisterFile.java:14]: Good catch, will update.

File: src/main/java/group1/mips_simulator/components/instructionParts/Instruction.java [hunk 8]
```diff
@@ -13,25 +10,18 @@ public class Instruction {
     public static final int BIT_COUNT = 16;
```
[the remark this answers is not in the document]
Comment (toon-leader-bacon) [reply src/main/java/group1/mips_simulator/components/instructionParts/Instruction.java:10]: Good catch, will udpate.

**Pull request merged**